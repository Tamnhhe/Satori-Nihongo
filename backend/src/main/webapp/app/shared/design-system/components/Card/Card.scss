@import '../../tokens/variables';

.satori-card {
  background: white;
  border-radius: $radius-lg;
  transition: all $transition-base;
  position: relative;
  overflow: hidden;

  // Padding variants
  &--padding-none {
    padding: 0;
  }

  &--padding-sm {
    padding: $spacing-4;
  }

  &--padding-md {
    padding: $spacing-6;
  }

  &--padding-lg {
    padding: $spacing-8;
  }

  &--padding-xl {
    padding: $spacing-10;
  }

  // Shadow variants
  &--shadow-none {
    box-shadow: none;
  }

  &--shadow-sm {
    box-shadow: $shadow-sm;
  }

  &--shadow-md {
    box-shadow: $shadow-md;
  }

  &--shadow-lg {
    box-shadow: $shadow-lg;
  }

  &--shadow-xl {
    box-shadow: $shadow-xl;
  }

  // Border
  &--border {
    border: 1px solid $gray-200;
  }

  // Hover effect
  &--hover {
    &:hover {
      box-shadow: $shadow-md;
      transform: translateY(-1px);
    }
  }

  // Clickable
  &--clickable {
    cursor: pointer;

    &:hover {
      box-shadow: $shadow-md;
    }

    &:active {
      transform: translateY(0);
    }
  }

  // Card components
  &__header {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin-bottom: $spacing-4;
    padding-bottom: $spacing-4;
    border-bottom: 1px solid $gray-200;

    &-content {
      flex: 1;

      h1,
      h2,
      h3,
      h4,
      h5,
      h6 {
        margin: 0;
        font-weight: $font-weight-semibold;
        color: $gray-900;
      }

      h1 {
        font-size: $text-2xl;
      }
      h2 {
        font-size: $text-xl;
      }
      h3 {
        font-size: $text-lg;
      }
      h4 {
        font-size: $text-base;
      }
      h5 {
        font-size: $text-sm;
      }
      h6 {
        font-size: $text-xs;
      }
    }

    &-actions {
      display: flex;
      align-items: center;
      gap: $spacing-2;
      margin-left: $spacing-4;
    }
  }

  &__body {
    color: $gray-700;
    line-height: $leading-relaxed;

    p {
      margin: 0 0 $spacing-4 0;

      &:last-child {
        margin-bottom: 0;
      }
    }
  }

  &__footer {
    display: flex;
    align-items: center;
    margin-top: $spacing-6;
    padding-top: $spacing-4;
    border-top: 1px solid $gray-200;

    &--left {
      justify-content: flex-start;
    }

    &--center {
      justify-content: center;
    }

    &--right {
      justify-content: flex-end;
    }

    &--between {
      justify-content: space-between;
    }
  }
}

// Card in grid layouts
.satori-card-grid {
  display: grid;
  gap: $spacing-6;

  &--cols-1 {
    grid-template-columns: 1fr;
  }

  &--cols-2 {
    grid-template-columns: repeat(2, 1fr);
  }

  &--cols-3 {
    grid-template-columns: repeat(3, 1fr);
  }

  &--cols-4 {
    grid-template-columns: repeat(4, 1fr);
  }

  // Responsive grid
  @media (max-width: $breakpoint-lg) {
    &--cols-4 {
      grid-template-columns: repeat(2, 1fr);
    }

    &--cols-3 {
      grid-template-columns: repeat(2, 1fr);
    }
  }

  @media (max-width: $breakpoint-md) {
    &--cols-4,
    &--cols-3,
    &--cols-2 {
      grid-template-columns: 1fr;
    }
  }
}
