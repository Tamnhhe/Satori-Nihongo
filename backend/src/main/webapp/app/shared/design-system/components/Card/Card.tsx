import React from 'react';
import { useKeyboardNavigation, useAriaAttributes } from '../../../hooks/useAccessibility';
import './Card.scss';

export interface CardProps {
  children: React.ReactNode;
  className?: string;
  padding?: 'none' | 'sm' | 'md' | 'lg' | 'xl';
  shadow?: 'none' | 'sm' | 'md' | 'lg' | 'xl';
  border?: boolean;
  hover?: boolean;
  onClick?: () => void;
  onKeyDown?: (event: React.KeyboardEvent) => void;
  as?: 'div' | 'article' | 'section';
  role?: string;
  'aria-label'?: string;
  'aria-labelledby'?: string;
  'aria-describedby'?: string;
  tabIndex?: number;
  id?: string;
}

export const Card: React.FC<CardProps> = ({
  children,
  className = '',
  padding = 'md',
  shadow = 'sm',
  border = true,
  hover = false,
  onClick,
  onKeyDown,
  as: Component = 'div',
  role,
  'aria-label': ariaLabel,
  'aria-labelledby': ariaLabelledBy,
  'aria-describedby': ariaDescribedBy,
  tabIndex,
  id,
}) => {
  const cardClasses = [
    'satori-card',
    `satori-card--padding-${padding}`,
    `satori-card--shadow-${shadow}`,
    border && 'satori-card--border',
    hover && 'satori-card--hover',
    onClick && 'satori-card--clickable',
    className,
  ]
    .filter(Boolean)
    .join(' ');

  // Enhanced keyboard navigation for clickable cards
  const { handleKeyDown: handleKeyboardNav } = useKeyboardNavigation(onClick ? () => onClick() : undefined);

  const handleKeyDownEvent = (event: React.KeyboardEvent) => {
    if (onKeyDown) {
      onKeyDown(event);
    }
    if (!event.defaultPrevented && onClick) {
      handleKeyboardNav(event.nativeEvent);
    }
  };

  // ARIA attributes
  const ariaAttributes = useAriaAttributes(role || (onClick ? 'button' : undefined), ariaLabel, ariaLabelledBy, ariaDescribedBy);

  const componentProps = {
    id,
    className: cardClasses,
    onClick,
    onKeyDown: onClick ? handleKeyDownEvent : onKeyDown,
    tabIndex: onClick ? (tabIndex ?? 0) : tabIndex,
    ...ariaAttributes,
  };

  return <Component {...componentProps}>{children}</Component>;
};

export interface CardHeaderProps {
  children: React.ReactNode;
  className?: string;
  actions?: React.ReactNode;
}

export const CardHeader: React.FC<CardHeaderProps> = ({ children, className = '', actions }) => {
  return (
    <div className={`satori-card__header ${className}`}>
      <div className="satori-card__header-content">{children}</div>
      {actions && <div className="satori-card__header-actions">{actions}</div>}
    </div>
  );
};

export interface CardBodyProps {
  children: React.ReactNode;
  className?: string;
}

export const CardBody: React.FC<CardBodyProps> = ({ children, className = '' }) => {
  return <div className={`satori-card__body ${className}`}>{children}</div>;
};

export interface CardFooterProps {
  children: React.ReactNode;
  className?: string;
  align?: 'left' | 'center' | 'right' | 'between';
}

export const CardFooter: React.FC<CardFooterProps> = ({ children, className = '', align = 'right' }) => {
  return <div className={`satori-card__footer satori-card__footer--${align} ${className}`}>{children}</div>;
};
