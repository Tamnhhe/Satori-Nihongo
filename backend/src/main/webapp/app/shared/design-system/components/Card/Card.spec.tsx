import React from 'react';
import { render, screen } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import { Card } from './Card';

expect.extend(toHaveNoViolations);

describe('Card Component', () => {
  it('should render with default props', () => {
    render(
      <Card>
        <p>Card content</p>
      </Card>,
    );

    const card = screen.getByTestId('card');
    expect(card).toBeInTheDocument();
    expect(card).toHaveClass('card');
    expect(screen.getByText('Card content')).toBeInTheDocument();
  });

  it('should render with title', () => {
    render(
      <Card title="Test Card">
        <p>Card content</p>
      </Card>,
    );

    expect(screen.getByText('Test Card')).toBeInTheDocument();
    expect(screen.getByTestId('card-header')).toBeInTheDocument();
  });

  it('should render with actions', () => {
    const actions = [<button key="edit">Edit</button>, <button key="delete">Delete</button>];

    render(
      <Card title="Test Card" actions={actions}>
        <p>Card content</p>
      </Card>,
    );

    expect(screen.getByTestId('card-actions')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /edit/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /delete/i })).toBeInTheDocument();
  });

  it('should apply different variants', () => {
    const { rerender } = render(<Card variant="elevated">Content</Card>);
    expect(screen.getByTestId('card')).toHaveClass('card-elevated');

    rerender(<Card variant="outlined">Content</Card>);
    expect(screen.getByTestId('card')).toHaveClass('card-outlined');

    rerender(<Card variant="filled">Content</Card>);
    expect(screen.getByTestId('card')).toHaveClass('card-filled');
  });

  it('should handle loading state', () => {
    render(
      <Card loading>
        <p>Content</p>
      </Card>,
    );

    expect(screen.getByTestId('card-skeleton')).toBeInTheDocument();
    expect(screen.queryByText('Content')).not.toBeInTheDocument();
  });

  it('should handle hover effects', () => {
    render(
      <Card hoverable>
        <p>Hoverable content</p>
      </Card>,
    );

    expect(screen.getByTestId('card')).toHaveClass('card-hoverable');
  });

  it('should render with custom className', () => {
    render(
      <Card className="custom-card">
        <p>Content</p>
      </Card>,
    );

    expect(screen.getByTestId('card')).toHaveClass('card', 'custom-card');
  });

  it('should render footer when provided', () => {
    render(
      <Card footer={<div>Footer content</div>}>
        <p>Main content</p>
      </Card>,
    );

    expect(screen.getByTestId('card-footer')).toBeInTheDocument();
    expect(screen.getByText('Footer content')).toBeInTheDocument();
  });

  it('should be accessible', async () => {
    const { container } = render(
      <Card title="Accessible Card" role="article" aria-labelledby="card-title">
        <p>This is accessible content</p>
      </Card>,
    );

    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should support keyboard navigation when interactive', () => {
    const handleClick = jest.fn();
    render(
      <Card onClick={handleClick} tabIndex={0}>
        <p>Interactive card</p>
      </Card>,
    );

    const card = screen.getByTestId('card');
    expect(card).toHaveAttribute('tabIndex', '0');
    expect(card).toHaveClass('card-interactive');
  });

  it('should match snapshot', () => {
    const { container } = render(
      <div>
        <Card title="Basic Card">
          <p>Basic content</p>
        </Card>
        <Card title="Advanced Card" variant="elevated" actions={[<button key="action">Action</button>]} footer={<span>Footer</span>}>
          <p>Advanced content</p>
        </Card>
        <Card loading />
      </div>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
