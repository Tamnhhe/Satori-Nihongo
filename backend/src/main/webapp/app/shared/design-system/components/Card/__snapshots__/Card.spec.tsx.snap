// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component should match snapshot 1`] = `
<div>
  <div
    class="satori-card satori-card--padding-md satori-card--shadow-sm satori-card--border"
  >
    <p>
      Basic content
    </p>
  </div>
  <div
    class="satori-card satori-card--padding-md satori-card--shadow-sm satori-card--border"
  >
    <p>
      Advanced content
    </p>
  </div>
  <div
    class="satori-card satori-card--padding-md satori-card--shadow-sm satori-card--border"
  />
</div>
`;
